### PR TITLE
KAFKA-15344: Streams task should cache consumer nextOffsets

### PR DESCRIPTION
**Pull request by aliehsaeedii**
This PR augments Streams messages with leader epoch. In case of empty buffer queues, the last offset and leader epoch are retrieved from the streams task 's cache of nextOffsets.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (aliehsaeedii) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: I am wondering if, instead of setting, an update should be done. Then, instead of overwriting the former map, we just update it by adding new entries or updating the existing ones.

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: This doesn't seem to achieve the goal that we are intending. Because the offset that we want to commit is exactly not `lastRecordOffset + 1`, but the position inside the consumer. I think we need to pass the OffsetAndMetadata into this class as part of the constructor.

That would mean we need to, after all, propose a KIP that extends the constructor of this class, because the constructor is public. Sorry, I missed that before. Let's implement it first, before we write the KIP, just to make sure that we get it right this time!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Would it be better to not add `setNextOffsetsAndMetadata` but make it part of `addRecord` to begin with?

In the end, we only need to pass in `nextRecord.offset` for the `partition` we are currently processing, plus a leader epoch, but not the full map of `nextRecords()`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java:193]:
```suggestion
        return headRecord == null ? Optional.empty() : headRecord.leaderEpoch();
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/SynchronizedPartitionGroup.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/SynchronizedPartitionGroup.java:76]: Why do we return `empty` instead of calling `wrapped...` like all other methods do?

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java
Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: Good callout!

Also wondering if `lastRecord.leaderEpoch()`is what we want? Or should it be `nextRecord.leaderEpoch()` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractPartitionGroup.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractPartitionGroup.java:60]: Would this really be an `Optional`? In the end, we _always_ need to have some leader epoch -- or do I misunderstand something?

Comment (aliehsaeedii) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractPartitionGroup.java:60]: leader epoch type is always `Optional<Integer>` (at least across all Consumer code)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (aliehsaeedii) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: I agree:)

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: Could be final

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: Who calls this constructor, is it in the Queues code? Do we really want to make `nextOffsets` null here? It seems better to initialize it to an empty map if no nextOffsets are available. It could be better to just have one constructor and pass an empty map for `nextOffsets` in the case of share consumers explicitly - it's easy to use this overload and "forget" to pass the right nextOffsets.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ClassicKafkaConsumer.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ClassicKafkaConsumer.java]: If both nextFetchOffsets and lastEpochs are coming from the fetch, could we just create `OffsetAndMetadata` directly inside a map `fetch.nextOffsetAndMetadata`? We'd avoid dealing with multiple maps this way it seems.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java:193]: Yeah we are distinguishing three cases here: `null` - queue empty. `Optional.empty()` - queue not empty, leader epoch not set on head record. otherwise, leader epoch. This is a bit confusing, but I think it's correct.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java]: Same here, I would be interested in when this constructor is called. I'm worried that we forget to pass the correct nextOffset information if this constructor is called.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (lucasbru) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Reminder to not approve this as long as we have commented out test cases.

**Review by lucasbru — COMMENTED** [review]
Hi @aliehsaeedii . I left some comments

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (lucasbru) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Not sure why you are creating a null pointer exception here?


Also, I wonder if we are sure that we will never hit this branch. In principle, we will get here whenever we try to commit offsets for a topic partition, for which we have not received any records. If I understand the code correctly, we should not get here, but honestly I'm not 100%.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (lucasbru) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: I think if you throw a fatal exception, you don't need to also log it. It should be logged by the code that catches the exception.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (lucasbru) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: It may be nice to get rid of this `instanceof` check, e.g. by extending the signature of `addRecords` with a map for `nextOffsets`, instead of using `updateNextOffsets`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (aliehsaeedii) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Yes, I also think that the next offset and leader epoch are existing in the map for sure. BTW, I was hesitating to retrieve a key from a map without checking the existence of the key. NullPointer refers to the fact that the key is not existing in the map. But I'm OK to remove it as well.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (aliehsaeedii) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Anyway, we need the `instanceof` check to call the newly defined `addRecords`!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (aliehsaeedii) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Oh, it was mistakenly commented out!

**Comment by aliehsaeedii** [comment]
> Hi @aliehsaeedii . I left some comments

Thanks. I think some of them must be applied in KIP's PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java:193]: Thanks. Time for an in-line comment.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: I meant we change the existing `addRecords` method (which is defined in the supertype).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Let's keep the branch. Throw an exception or at least log an error, so we detect if our assumption is incorrect.

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java
Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: The position inside the consumer is `nextFetchPosition` with leader epoch `lastEpoch`, so I think we want to commit that.

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java
Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: I agree with initializing the `nextOffsets` with an empty map in the current constructor but changing the current constructor input parameters may break the users code.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java:193]: @alhsaeedi I think @mjsax suggested adding an inline comment explaining this. Could you please add it?

**Review by lucasbru — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
